### DISCUSSION TITLE
Cannot get debugger working on Windows - won't map to target or step through the listings

### DISCUSSION THREAD

**Discussion by smerriman** [NationalSecurityAgency/ghidra#5689] [Debugger]
I am working through the termmines tutorial in Ghidra's documentation and cannot get the debugger to work no matter what I try.

After creating a project and importing termmines.exe, I open the debugger, analyze the file (this shows me the static listing and decompiled code fine), and then connect to dbgeng locally IM-VM. I first get an error about the mapping not existing:

![image](https://github.com/NationalSecurityAgency/ghidra/assets/6667004/5159d7cd-5be6-42cb-8998-7e944289f2d1)

Have hunted through some past threads on this and saw recommendations about finding the process in the modules window, but there's only one thing in there and if I try mapping it, it doesn't help:

![image](https://github.com/NationalSecurityAgency/ghidra/assets/6667004/ae1d8260-a12c-4f82-b5b1-ee11e16909a1)

If I instead hit f5 / press the resume button in the Objects window, more things appear in the interpreter:

![image](https://github.com/NationalSecurityAgency/ghidra/assets/6667004/b6ff93d5-ed2b-477d-a118-dfad832f3b1b)

And then f5 again, a lot more:

![image](https://github.com/NationalSecurityAgency/ghidra/assets/6667004/dc12db59-fe79-4c62-a295-d4094a40b5a4)

Only now does termmines.exe open in a separate command prompt. While I can play the game, the resume / step into etc buttons are now all greyed out in the Objects window, and the dynamic window is showing some contents of ntdll.dll, so I can't use the debugger at all to see what's going on in the program itself.

Still nothing that looks like termmines.exe in the modules window either:

![image](https://github.com/NationalSecurityAgency/ghidra/assets/6667004/0b38b24c-9ca8-4333-a7af-b74c4384329a)

I feel like I must be missing something fundamental, and have been going through all past support threads, but haven't been able to find anything that has helped.

**Comment by smerriman** [comment]
OK, I think I might have finally been able to figure out my problem.

You do indeed need to map that image000... to the exe. And you'll also need to manually place a breakpoint at the start of the program's execution. And lastly, after you start the execution, you have to go back to that breakpoint and double click it to turn it on, since *every single time* you run it, it starts off in a disabled state.

A bit of a mess but hopefully this helps someone else..

**Comment by smerriman** [comment]
OK, a follow up question - while it's now stepping through fine, I'm unable to see the values of the majority of variables, as they're throwing IllegalArgumentExceptions:

![image](https://github.com/NationalSecurityAgency/ghidra/assets/6667004/04e5629f-2c72-402f-bb3c-5f6de613cb8b)

Any suggestions?

**Reply by astrelsky** [reply]
@d-millar you might want to take a look at this.

**Comment by d-millar** [comment]
I think this might be one for @nsadeveloper789 - he designed this feature and was fixing related issues recently.  @smerriman - what version of Ghidra are you running?  I ask as I'm not seeing this behavior in master - will check 10.3.2, but am pretty sure there's already a fix in for this.

**Reply by smerriman** [reply]
I'm using 10.3.2.

**Comment by Nemoumbra** [comment]
I've got some questions too about this whole debugging stuff Ghidra can do. Excuse me, but what do you think of this error message?

![image](https://github.com/NationalSecurityAgency/ghidra/assets/62447396/3c188522-23dc-4b49-9784-a90defe8303c)

The context: I'm exploring the possible debugging scenarios in my Windows Ghidra 10.3.2 and the first one is the `gdb`:

![image](https://github.com/NationalSecurityAgency/ghidra/assets/62447396/58defdc9-12cf-4538-a440-7748b2031da9)

I enter my MinGW gdb path there and get this useless error message. Then I replace `gdb.exe` with `lol.exe` and the message remains the same. I highly doubt that's fine. Please tell me, what am I doing wrong?

**Comment by d-millar** [comment]
@Nemoumbra Hitting that "Details >>>" button and sending us the stack trace would be very very helpful.

**Reply by d-millar** [reply]
Also, this probably should be a new ticket as it does not appear to have anything to do with @smerriman's issue.

**Reply by Nemoumbra** [reply]
I suppose you're right. There is nothing interesting about the Details (and that's why I didn't include that info).
```
com.sun.jna.LastErrorException: GetLastError() returned 2
---------------------------------------------------
Build Date: 2023-Jul-11 1640 EDT
Ghidra Version: 10.3.2
Java Home: k:\Shared_storage\Data\JDK19\jdk-19.0.2
JVM Version: Oracle Corporation 19.0.2
OS: Windows 10 10.0 amd64
```

**Reply by Nemoumbra** [reply]
https://github.com/NationalSecurityAgency/ghidra/discussions/5698
@d-millar

**Comment by d-millar** [comment]
Wow, ok - that is spectacularly uninformative.  :)

The immediate thing that confuses me is the JNA error, given our implementation of the gdb agent doesn't use JNA directly.  Any chance you could walk me through the exact steps you took?  I.e. did Targets->Connect throw this error immediately?  One thing to try for sure - try using double-slashes in your path (Java is not a fan of the backslash).

**Comment by d-millar** [comment]
@smerriman OK, spoke with @nsadeveloper789 - he confirmed this is a know issue.  Basically, the fact that you're running a WOW target means that 32-bit registers are not being identified correctly in a 64-bit space.  We have a ticket in to fix it - will keep you posted.
